VariantShredding.md

```diff
@@ -25,276 +25,302 @@
- optional binary variant_value;
- optional group object {
-  optional group a {
-   optional binary variant_value;
-   optional int64 typed_value;
-  }
-  optional group b {
-   optional binary variant_value;
-   optional group object {
-    optional group c {
-      optional binary variant_value;
-      optional binary typed_value (STRING);
-    }
-   }
-  }
- }
+required group measurement (VARIANT) {
+  required binary metadata;
+  optional binary value;
+  optional int64 typed_value;
 }
 ```
 
-| Variant Value | Top-level variant_value | b.variant_value | a.typed_value | a.variant_value | b.object.c.typed_value | b.object.c.variant_value | Notes | 
-|---------------|-------------------------|-----------------|---------------|-----------------|------------------------|--------------------------|-------|
-| {a: 123, b: {c: “hello”}} | null | null | 123 | null | hello | null | All values shredded |
-| {a: 1.23, b: {c: “123”}} | null | null | null | 1.23 | 123 | null | a is not an integer |
-| {a: 123, b: {c: null}} | null | null | null | 123 | null | null | b.object.c set to non-null to indicate VariantNull |
-| {a: 123, b: {} | null | null | null | 123 | null | null | b.object.c set to null, to indicate that c is missing |
-| {a: 123, d: 456} | {d: 456} | null | 123 | null | null | null | Extra field d is stored as variant_value |
-| [{a: 1, b: {c: 2}}, {a: 3, b: {c: 4}}] | [{a: 1, b: {c: 2}}, {a: 3, b: {c: 4}}] | null | null | null | null | null | Not an object |
+The series of measurements `34, null, "n/a", 100` would be stored as:
 
-# Parquet Layout
+| Value   | `metadata`       | `value`               | `typed_value` |
+|---------|------------------|-----------------------|---------------|
+| 34      | `01 00` v1/empty | null                  | `34`          |
+| null    | `01 00` v1/empty | `00` (null)           | null          |
+| "n/a"   | `01 00` v1/empty | `13 6E 2F 61` (`n/a`) | null          |
+| 100     | `01 00` v1/empty | null                  | `100`         |
 
-The `array` and `object` fields represent Variant array and object types, respectively.
-Arrays must use the three-level list structure described in https://github.com/apache/parquet-format/blob/master/LogicalTypes.md.
+Both `value` and `typed_value` are optional fields used together to encode a single value.
+Values in the two fields must be interpreted according to the following table:
 
-An `object` field must be a group.
-Each field name of this inner group corresponds to the Variant value's object field name.
-Each inner field's type is a recursively shredded variant value: that is, the fields of each object field must be one or more of `object`, `array`, `typed_value` or `variant_value`.
+| `value`  | `typed_value` | Meaning                                                  |
+|----------|---------------|----------------------------------------------------------|
+| null     | null          | The value is missing                                     |
```

Comment:
Just to be clear, this is only allowed for object fields, right? You mention in the array section that array elements must have one of them non-null, but I think that's also true for the top-level `value`/`typed_value`, right?

Reply:
I think this would be good to clarify. I think that we could state that the variant could be null this way.

Reply:
This would be Variant null (i.e. present but null)? I guess this is the same as the both-null case for array elements, which still seems to me more like an error state, but I guess Variant null is the best option if a reader doesn't want to fail.

Reply:
My understanding was both typed_value and value being null meant variant being null if this is for the top level variant and field not being present for a shredded field.

Reply:
My preference would be to make it illegal for the top level variant field to be variant-null (same for array elements). It seems like it adds a relatively rare special case that readers would need to handle, and doesn't add much value, since the null can be encoded in `value` just like for shredded fields. I don't feel too strongly if the consensus that it adds value.

Reply:
@cashmand, the problem with "make it illegal" is that there is no functional way to do this. We can state that writers should not set both `value` and `typed_value` to null in certain cases, but we have to define what to do if there is data that is actually written that way in order to have consistent and reliable behavior.

That's why this is called out as a requirement:
> If a Variant is missing in a context where a value is required, readers must either fail or return a Variant null: basic type 0 (primitive) and physical type 0 (null).